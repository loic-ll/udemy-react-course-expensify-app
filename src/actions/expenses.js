import {database} from '../firebase/firebase';

const expensesRef = database.ref('expenses');

const defaultValues = {
  amount: 0,
  createdAt: 0,
  description: '',
  note: '',
};

// ADD_EXPENSE
export const addExpense = expense => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense = (expense = defaultValues) => {
  return dispatch => {
    return expensesRef.push(expense).then(ref => {
      dispatch(
        addExpense({
          id: ref.key,
          ...expense,
        }),
      );
    });
  };
};

// EDIT_EXPENSE
export const editExpense = (id, values) => ({
  type: 'EDIT_EXPENSE',
  id,
  values,
});

export const startEditExpense = (id, values) => {
  return dispatch => {
    return expensesRef
      .child(id)
      .update(values)
      .then(() => {
        dispatch(editExpense(id, values));
      });
  };
};

// REMOVE_EXPENSE
export const removeExpense = id => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const startRemoveExpense = id => {
  return dispatch => {
    return expensesRef
      .child(id)
      .remove()
      .then(() => {
        dispatch(removeExpense(id));
      });
  };
};

// SET_EXPENSES

export const setExpenses = expenses => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => {
  return dispatch => {
    return expensesRef.once('value').then(snapshot => {
      const expenses = [];
      snapshot.forEach(childsnap => {
        expenses.push({
          id: childsnap.key,
          ...childsnap.val(),
        });
      });

      dispatch(setExpenses(expenses));
    });
  };
};
