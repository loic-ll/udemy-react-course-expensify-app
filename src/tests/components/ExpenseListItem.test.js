import React from 'react'
import { shallow } from 'enzyme'
import ExpenseListItem  from '../../components/ExpenseListItem'
import expenses from '../fixtures/expenses'

test('ExpenseListItem should render without any props', () => {
  const wrapper = shallow(<ExpenseListItem />);
  expect(wrapper).toMatchSnapshot();
});

test('ExpenseListItem should render with props', () => {
  const wrapper = shallow(<ExpenseListItem {...expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});
