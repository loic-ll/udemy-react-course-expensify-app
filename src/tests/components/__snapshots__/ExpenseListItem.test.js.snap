// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseListItem should render with props 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/1"
  >
    <h3>
      bcd
    </h3>
  </Link>
  <p>
    1
    -
    -345600000
  </p>
</div>
`;

exports[`ExpenseListItem should render without any props 1`] = `
<div>
  <Link
    replace={false}
    to="/edit/undefined"
  >
    <h3 />
  </Link>
  <p>
    -
  </p>
</div>
`;
